content/blog/2023-04-04-the-pci-device-that-wasnt.md

```diff
@@ -0,0 +1,289 @@
+
+
+# The bad access
+
+```c
+static bool vga_arbiter_add_pci_device(struct pci_dev *pdev)
+{
+        struct vga_device *vgadev;
+        unsigned long flags;
+        struct pci_bus *bus;
+        struct pci_dev *bridge;
+        u16 cmd;
+
+        /* Only deal with VGA class devices */
+        if ((pdev->class >> 8) != PCI_CLASS_DISPLAY_VGA)
+                return false;
+```
+
+We're blowing up on the read to `pdev->class`, and it's not something like the data being uninitialised, it's out-of-bounds.  If we look back at the call trace:
+
+```text
+vga_arbiter_add_pci_device+0x60/0xe00
+pci_notify+0x88/0x444
+notifier_call_chain+0x104/0x320
+blocking_notifier_call_chain+0xa0/0x140
+device_add+0xac8/0x1d30
+device_register+0x58/0x80
+vio_register_device_node+0x9ac/0xce0
+vio_bus_scan_register_devices+0xc4/0x13c
+```
+
+This thing is a VIO device, not a PCI device!  Let's jump into the caller, `pci_notify()`, to find out how we got our `pdev`.
+
+```c
+static int pci_notify(struct notifier_block *nb, unsigned long action,
+                      void *data)
+{
+        struct device *dev = data;
+        struct pci_dev *pdev = to_pci_dev(dev);
+```
+
+So `pci_notify()` gets called with our VIO device (somehow), and we're converting that `struct device` into a `struct pci_dev` with no error checking.  We could solve this particular bug by just checking that our device is *actually* a PCI device before we proceed - but we're in a function called `pci_notify`, we're expecting a PCI device to come in, so this would just be a bandaid.
+
+`to_pci_dev()` works like other struct containers in the kernel - `struct pci_dev` contains a `struct device` as a member, so the `container_of()` function returns an address based on where a `struct pci_dev` would have to be if the given `struct device` was actually a PCI device.  Since we know it's not actually a PCI device and this `struct device` does not actually sit inside a `struct pci_dev`, our `pdev` is now pointing to some random place in memory, hence our access to a member like `class` is caught by KASAN.
+
+Now we know why and how we're blowing up, but we still don't understand how we got here, so let's back up further.
+
+
+# Notifiers
+
+The kernel's device subsystem allows consumers to register callbacks to be notified on given events.  I'm not going to go into a ton of detail on how they work, because I don't fully understand myself, and there's a lot of internals of the device subsystem involved.
```

Comment:
(Pedantic) consider:
'The kernel's device subystem allows consumes to register callbacks so that they can be notified of a given event.'
Just ensures that the object of the notifying is the consumer in this sentence.

Reply:
> Just ensures that the object of the notifying is the consumer in this sentence.

I like your funny words, magic man